superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx

```diff
@@ -0,0 +1,1307 @@
+
+    let result;
+
+    // Cycle through source options to find the selected option
+    sourceOptions.forEach(source => {
+      if (source.value === database.value || source.value === database.id) {
+        result = source;
+      }
+    });
+
+    return result;
+  };
+
+  const loadDashboardOptions = (input = '') => {
+    const query = rison.encode({ filter: input });
+    return SupersetClient.get({
+      endpoint: `/api/v1/dashboard?q=${query}`,
+    }).then(
+      response => {
+        const list = response.json.result.map((item: any) => ({
+          value: item.id,
+          label: item.dashboard_title,
+        }));
+
+        setDashboardOptions(list);
+
+        // Find source if current alert has one set
+        if (
+          currentAlert &&
+          currentAlert.dashboard &&
+          !currentAlert.dashboard.label
+        ) {
+          updateAlertState('dashboard', getDashboardData());
+        }
+
+        return list;
+      },
+      badResponse => {
+        return [];
+      },
+    );
+  };
+
+  const getDashboardData = (db?: MetaObject) => {
+    const dashboard = db || currentAlert?.dashboard;
+
+    if (!dashboard || dashboard.label) {
+      return null;
+    }
+
+    let result;
+
+    // Cycle through dashboard options to find the selected option
+    dashboardOptions.forEach(dash => {
+      if (dash.value === dashboard.value || dash.value === dashboard.id) {
+        result = dash;
+      }
+    });
+
+    return result;
+  };
+
+  const loadChartOptions = (input = '') => {
+    const query = rison.encode({ filter: input });
+    return SupersetClient.get({
+      endpoint: `/api/v1/chart?q=${query}`,
+    }).then(
+      response => {
+        const list = response.json.result.map((item: any) => ({
+          value: item.id,
+          label: item.slice_name,
+        }));
+
+        setChartOptions(list);
+
+        // Find source if current alert has one set
+        if (currentAlert && currentAlert.chart && !currentAlert.chart.label) {
+          updateAlertState('chart', getChartData());
+        }
+
+        return list;
+      },
+      badResponse => {
+        return [];
+      },
+    );
+  };
+
+  const getChartData = (chartData?: MetaObject) => {
+    const chart = chartData || currentAlert?.chart;
+
+    if (!chart || chart.label) {
+      return null;
+    }
+
+    let result;
+
+    // Cycle through chart options to find the selected option
+    chartOptions.forEach(slice => {
+      if (slice.value === chart.value || slice.value === chart.id) {
+        result = slice;
+      }
+    });
+
+    return result;
+  };
+
+  // Updating alert/report state
+  const updateAlertState = (name: string, value: any) => {
+    const data = {
+      ...currentAlert,
+    };
+
+    data[name] = value;
+    setCurrentAlert(data);
+  };
+
+  // Handle input/textarea updates
+  const onTextChange = (
+    event:
+      | React.ChangeEvent<HTMLTextAreaElement>
+      | React.ChangeEvent<HTMLInputElement>,
```

Comment:
could also probably do `React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>` to be more terse